openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -436,15 +436,18 @@ def create_library(
     )
     # Now create the library reference in our database:
     try:
-        ref = ContentLibrary.objects.create(
-            org=org,
-            slug=slug,
-            type=library_type,
-            bundle_uuid=bundle.uuid,
-            allow_public_learning=allow_public_learning,
-            allow_public_read=allow_public_read,
-            license=library_license,
-        )
+        # Atomic transaction required because if this fails,
+        # we need to delete the bundle in the exception handler.
+        with transaction.atomic():
```

Comment:
I was hitting this issue when running the tests locally. If the content library already exists, we proceed to the exception handler, but it seems that the failed transaction isn't closed when the exception is thrown. So it wouldn't let us delete the bundle.

Reply:
Probably, however: the line that fails without this change is *within* [`ContentLibrariesTest::test_library_validation`](https://github.com/open-craft/edx-platform/blob/jill/TNL-8746/openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py#L193), and since using `TransactionTestCase` only causes the database to be reset *between* tests, changing the base test class doesn't fix the problem.

For the record, I tried this diff, and still the test failed:

```bash
root@studio:/edx/app/edxapp/edx-platform$ python -Wd -m pytest --ds=cms.envs.test openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py::ContentLibrariesTest::test_library_validation
```

```diff
diff --git a/openedx/core/djangoapps/content_libraries/api.py b/openedx/core/djangoapps/content_libraries/api.py
index c1d95c9d68..0d9ee8c899 100644
--- a/openedx/core/djangoapps/content_libraries/api.py
+++ b/openedx/core/djangoapps/content_libraries/api.py
@@ -65,7 +65,7 @@ from django.conf import settings
 from django.contrib.auth.models import AbstractUser, Group
 from django.core.exceptions import PermissionDenied
 from django.core.validators import validate_unicode_slug
-from django.db import IntegrityError, transaction
+from django.db import IntegrityError
 from django.utils.translation import gettext as _
 from elasticsearch.exceptions import ConnectionError as ElasticConnectionError
 from lxml import etree
@@ -438,16 +438,15 @@ def create_library(
     try:
         # Atomic transaction required because if this fails,
         # we need to delete the bundle in the exception handler.
-        with transaction.atomic():
-            ref = ContentLibrary.objects.create(
-                org=org,
-                slug=slug,
-                type=library_type,
-                bundle_uuid=bundle.uuid,
-                allow_public_learning=allow_public_learning,
-                allow_public_read=allow_public_read,
-                license=library_license,
-            )
+        ref = ContentLibrary.objects.create(
+            org=org,
+            slug=slug,
+            type=library_type,
+            bundle_uuid=bundle.uuid,
+            allow_public_learning=allow_public_learning,
+            allow_public_read=allow_public_read,
+            license=library_license,
+        )
     except IntegrityError:
         delete_bundle(bundle.uuid)
         raise LibraryAlreadyExists(slug)  # lint-amnesty, pylint: disable=raise-missing-from
diff --git a/openedx/core/djangoapps/content_libraries/tests/base.py b/openedx/core/djangoapps/content_libraries/tests/base.py
index 34aee552ee..f5124ae304 100644
--- a/openedx/core/djangoapps/content_libraries/tests/base.py
+++ b/openedx/core/djangoapps/content_libraries/tests/base.py
@@ -12,7 +12,7 @@ from django.test import LiveServerTestCase
 from django.test.client import RequestFactory
 from django.test.utils import override_settings
 from organizations.models import Organization
-from rest_framework.test import APITestCase, APIClient
+from rest_framework.test import APITestCase, APITransactionTestCase, APIClient
 from search.search_engine_base import SearchEngine

 from common.djangoapps.student.tests.factories import UserFactory
@@ -398,7 +398,7 @@ class ContentLibrariesRestApiBlockstoreServiceTest(_ContentLibrariesRestApiTestM
 class ContentLibrariesRestApiTest(
     _ContentLibrariesRestApiTestMixin,
     BlockstoreAppTestMixin,
-    APITestCase,
+    APITransactionTestCase,
     LiveServerTestCase,   # LiveServerTestCase's base class is TransactionTestCase
 ):
     """
```

Reply:
Got it.